pkg/core/resources/model/marshal.go

```diff
@@ -0,0 +1,61 @@
+package model
+
+import (
+	"encoding/json"
+
+	"github.com/ghodss/yaml"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	util_proto "github.com/kumahq/kuma/pkg/util/proto"
+)
+
+var ToJSON = &marshal[[]byte]{protoFn: util_proto.ToJSON, fn: json.Marshal}
+var ToYAML = &marshal[[]byte]{protoFn: util_proto.ToYAML, fn: yaml.Marshal}
+var ToAny = &marshal[*anypb.Any]{
+	protoFn: util_proto.MarshalAnyDeterministic,
+	fn: func(a any) (*anypb.Any, error) {
+		bytes, err := json.Marshal(a)
+		if err != nil {
+			return nil, err
+		}
+		return &anypb.Any{
+			Value: bytes,
+		}, nil
+	},
+}
+
+var FromJSON = &unmarshal[[]byte]{protoFn: util_proto.FromJSON, fn: json.Unmarshal}
+var FromYAML = &unmarshal[[]byte]{protoFn: util_proto.FromYAML, fn: yaml.Unmarshal}
+var FromAny = &unmarshal[*anypb.Any]{
+	protoFn: util_proto.UnmarshalAnyTo,
+	fn: func(a *anypb.Any, dst any) error {
+		return json.Unmarshal(a.Value, dst)
+	},
+}
+
+type marshal[T any] struct {
+	protoFn func(proto.Message) (T, error)
+	fn      func(any) (T, error)
+}
+
+func (m *marshal[T]) ResourceSpec(desc ResourceTypeDescriptor, spec ResourceSpec) (T, error) {
+	if desc.IsPluginOriginated {
+		return m.fn(spec)
+	} else {
+		return m.protoFn(spec)
+	}
+}
+
+type unmarshal[T any] struct {
+	protoFn func(T, proto.Message) error
+	fn      func(T, any) error
+}
+
+func (m *unmarshal[T]) ResourceSpec(desc ResourceTypeDescriptor, src T, spec ResourceSpec) error {
+	if desc.IsPluginOriginated {
+		return m.fn(src, spec)
+	} else {
+		return m.protoFn(src, spec)
+	}
+}
```

Comment:
I feel like plain old functions would be more readable.
```
ToJson(desc ResourceTypeDescriptor, spec ResourceSpec) ([]byte, error)
FromJson(desc ResourceTypeDescriptor, src []byte, spec ResourceSpec) error
...
```

Or even with generics, but I don't understand why we create immutable objects with var with single function. Do we replace those vars somewhere?

Reply:
I had in mind creating different methods, not just `ResourceSpec(...)`, but for example:
```go
func (m *unmarshal[T]) ResourceSpecTo(desc ResourceTypeDescriptor, src []byte, spec ResourceSpec) error
func (m *unmarshal[T]) ResourceSpec(desc ResourceTypeDescriptor, src []byte) (ResourceSpec, error)
func (m *unmarshal[T]) ResourceTo(src []byte, r Resource) error
func (m *unmarshal[T]) ResourceTo(src []byte) (Resource, error)
```
But I think it's fine as it is with a single method. I'll replace immutable object with:
```go
func FromJSON(desc ResourceTypeDescriptor, src []byte, spec ResourceSpec) error {
	u := &unmarshal[[]byte]{protoFn: util_proto.FromJSON, fn: json.Unmarshal}
	return u.ResourceSpec(desc, src, spec)
}

func FromYAML(desc ResourceTypeDescriptor, src []byte, spec ResourceSpec) error {
	u := &unmarshal[[]byte]{protoFn: util_proto.FromYAML, fn: yaml.Unmarshal}
	return u.ResourceSpec(desc, src, spec)
}

func FromAny(desc ResourceTypeDescriptor, src *anypb.Any, spec ResourceSpec) error {
	u := &unmarshal[*anypb.Any]{
		protoFn: util_proto.UnmarshalAnyTo,
		fn: func(a *anypb.Any, dst any) error {
			return json.Unmarshal(a.Value, dst)
		},
	}
	return u.ResourceSpec(desc, src, spec)
}
```